File: packages/septr-node/src/__tests__/missing-auth.test.ts
import { describe, it, expect } from "vitest"
import { detectMissingAuth } from "../core/missing-auth"

describe("detectMissingAuth", () => {
  it("flags unprotected routes with no auth header", () => {
    const result = detectMissingAuth("/api/users", "GET", undefined)
    expect(result).not.toBeNull()
    expect(result!.type).toBe("missing_auth")
    expect(result!.severity).toBe("high")
    expect(result!.route).toBe("/api/users")
    expect(result!.method).toBe("GET")
  })

  it("flags unprotected routes with empty auth header", () => {
    const result = detectMissingAuth("/api/projects", "POST", "")
    expect(result).not.toBeNull()
    expect(result!.type).toBe("missing_auth")
  })

  it("flags routes with non-Bearer auth header", () => {
    const result = detectMissingAuth("/api/admin", "DELETE", "Basic dXNlcjpwYXNz")
    expect(result).not.toBeNull()
  })

  it("allows routes with Bearer token", () => {
    const result = detectMissingAuth("/api/users", "GET", "Bearer eyJhbGciOiJIUzI1NiJ9." + "eyJzdWIiOiIxMjMifQ")
    expect(result).toBeNull()
  })

  it("allows routes with Bearer token lowercase", () => {
    const result = detectMissingAuth("/api/data", "GET", "bearer some_token")
    expect(result).toBeNull()
  })

  it("allows public routes without auth", () => {
    const publicRoutes = [
      "/auth/login",
      "/login",
      "/register",
      "/signup",
      "/health",
      "/__septr_ping",
      "/favicon.ico",
    ]
    for (const route of publicRoutes) {
      expect(detectMissingAuth(route, "GET", undefined)).toBeNull()
    }
  })

  it("allows public routes case-insensitively", () => {
    expect(detectMissingAuth("/Auth/Login", "GET", undefined)).toBeNull()
    expect(detectMissingAuth("/HEALTH", "GET", undefined)).toBeNull()
  })

  it("flags nested API routes", () => {
    expect(detectMissingAuth("/api/v1/users/123/invoices", "GET", undefined)).not.toBeNull()
    expect(detectMissingAuth("/api/v1/organizations/teams", "POST", undefined)).not.toBeNull()
  })

  it("flags routes with arbitrary methods", () => {
    expect(detectMissingAuth("/api/data", "PATCH", undefined)).not.toBeNull()
    expect(detectMissingAuth("/api/data", "PUT", undefined)).not.toBeNull()
    expect(detectMissingAuth("/api/data", "DELETE", undefined)).not.toBeNull()
  })
})

describe("detectMissingAuth skip rules", () => {
  it("does not flag OPTIONS preflight requests", () => {
    expect(detectMissingAuth("/api/users", "OPTIONS", undefined)).toBeNull()
  })

  it("does not flag HEAD requests", () => {
    expect(detectMissingAuth("/api/users", "HEAD", undefined)).toBeNull()
  })

  it("does not flag static assets", () => {
    expect(detectMissingAuth("/static/main.js", "GET", undefined)).toBeNull()
    expect(detectMissingAuth("/app.css", "GET", undefined)).toBeNull()
  })

  it("still flags real API routes without auth", () => {
    expect(detectMissingAuth("/api/users", "GET", undefined)).not.toBeNull()
  })
})

File: packages/septr-go/missing_auth_test.go
package septr

import (
	"testing"
)

func TestDetectMissingAuth_NoHeader(t *testing.T) {
	ev := detectMissingAuth("/api/users", "GET", "")
	if ev == nil {
		t.Fatal("expected missing auth detection on unprotected route")
	}
	if ev.PatternID != "missing-auth-no-header" {
		t.Errorf("expected missing-auth-no-header, got %s", ev.PatternID)
	}
}

func TestDetectMissingAuth_WithBearer(t *testing.T) {
	ev := detectMissingAuth("/api/users", "GET", "Bearer token123")
	if ev != nil {
		t.Error("expected no detection when Bearer token present")
	}
}

func TestDetectMissingAuth_PublicRoutes(t *testing.T) {
	routes := []string{"/auth/signup", "/login", "/register", "/health", "/__septr_ping", "/favicon.ico"}
	for _, route := range routes {
		ev := detectMissingAuth(route, "GET", "")
		if ev != nil {
			t.Errorf("expected no detection on public route %s, got event", route)
		}
	}
}

func TestDetectMissingAuth_PublicRoutePrefix(t *testing.T) {
	ev := detectMissingAuth("/auth/callback", "POST", "")
	if ev != nil {
		t.Error("expected no detection on /auth/* path")
	}
}

func TestDetectMissingAuth_UnauthenticatedAdmin(t *testing.T) {
	ev := detectMissingAuth("/api/admin/users", "DELETE", "")
	if ev == nil {
		t.Fatal("expected missing auth detection on admin route without auth")
	}
	if ev.Severity != "high" {
		t.Errorf("expected severity high, got %s", ev.Severity)
	}
}

func TestMissingAuthSkipsOptionsPreflight(t *testing.T) {
	if ev := detectMissingAuth("/api/users", "OPTIONS", ""); ev != nil {
		t.Fatal("OPTIONS preflight should not be flagged")
	}
}

func TestMissingAuthSkipsHead(t *testing.T) {
	if ev := detectMissingAuth("/api/users", "HEAD", ""); ev != nil {
		t.Fatal("HEAD should not be flagged")
	}
}

func TestMissingAuthSkipsStaticAssets(t *testing.T) {
	if ev := detectMissingAuth("/static/main.js", "GET", ""); ev != nil {
		t.Fatal("static asset should not be flagged")
	}
}

func TestMissingAuthStillFlagsApiRoute(t *testing.T) {
	if ev := detectMissingAuth("/api/users", "GET", ""); ev == nil {
		t.Fatal("API route without auth should be flagged")
	}
}
